# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/cache/QuotaClient.cpp
# Commit: fbfdd8ee1b42
# Full Hash: fbfdd8ee1b42fb2d863b1d669a6047087d7e0595
# Author: Tom Tung <shes050117@gmail.com>
# Date: 2019-09-28 09:46:05
# Regressor Bug: 1581067
# File Overlap Count: 1
# Description:
#   Bug 1581067 - P1 - Remove the emptry body directory when removing files inside; r=asuth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D46586
# ==============================================================================

diff -r 2416d17d33c1 -r fbfdd8ee1b42 dom/cache/QuotaClient.cpp
--- a/dom/cache/QuotaClient.cpp	Fri Sep 27 15:26:14 2019 +0000
+++ b/dom/cache/QuotaClient.cpp	Fri Sep 27 16:54:22 2019 +0000
@@ -36,45 +36,91 @@
 using mozilla::dom::quota::UsageInfo;
 using mozilla::ipc::AssertIsOnBackgroundThread;
 
-static nsresult GetBodyUsage(nsIFile* aDir, const Atomic<bool>& aCanceled,
+static nsresult GetBodyUsage(nsIFile* aMorgueDir, const Atomic<bool>& aCanceled,
                              UsageInfo* aUsageInfo) {
   AssertIsOnIOThread();
 
   nsCOMPtr<nsIDirectoryEnumerator> entries;
-  nsresult rv = aDir->GetDirectoryEntries(getter_AddRefs(entries));
+  nsresult rv = aMorgueDir->GetDirectoryEntries(getter_AddRefs(entries));
   if (NS_WARN_IF(NS_FAILED(rv))) {
     return rv;
   }
 
-  nsCOMPtr<nsIFile> file;
-  while (NS_SUCCEEDED(rv = entries->GetNextFile(getter_AddRefs(file))) &&
-         file && !aCanceled) {
+  nsCOMPtr<nsIFile> bodyDir;
+  while (NS_SUCCEEDED(rv = entries->GetNextFile(getter_AddRefs(bodyDir))) &&
+         bodyDir && !aCanceled) {
     if (NS_WARN_IF(QuotaManager::IsShuttingDown())) {
       return NS_ERROR_ABORT;
     }
+    bool isDir;
+    rv = bodyDir->IsDirectory(&isDir);
+    if (NS_WARN_IF(NS_FAILED(rv))) {
+      return rv;
+    }
 
-    bool isDir;
-    rv = file->IsDirectory(&isDir);
+    if (!isDir) {
+      QuotaInfo dummy;
+      mozilla::DebugOnly<nsresult> result =
+          RemoveNsIFile(dummy, bodyDir, /* aTrackQuota */ false);
+      // Try to remove the unexpected files, and keep moving on even if it fails
+      // because it might be created by virus or the operation system
+      MOZ_ASSERT(NS_SUCCEEDED(result));
+      continue;
+    }
+
+    nsCOMPtr<nsIDirectoryEnumerator> subEntries;
+    rv = bodyDir->GetDirectoryEntries(getter_AddRefs(subEntries));
     if (NS_WARN_IF(NS_FAILED(rv))) {
       return rv;
     }
 
-    if (isDir) {
-      rv = GetBodyUsage(file, aCanceled, aUsageInfo);
+    bool isEmpty = true;
+    nsCOMPtr<nsIFile> bodyFile;
+    while (
+        NS_SUCCEEDED(rv = subEntries->GetNextFile(getter_AddRefs(bodyFile))) &&
+        bodyFile && !aCanceled) {
+      bool isDirectory;
+      rv = bodyFile->IsDirectory(&isDirectory);
       if (NS_WARN_IF(NS_FAILED(rv))) {
         return rv;
       }
-      continue;
-    }
+
+      if (isDirectory) {
+        QuotaInfo dummy;
+        mozilla::DebugOnly<nsresult> result =
+            RemoveNsIFileRecursively(dummy, bodyFile, /* aTrackQuota */ false);
+        // Try to remove the unexpected files, and keep moving on even if it
+        // fails because it might be created by virus or the operation system
+        MOZ_ASSERT(NS_SUCCEEDED(result));
+        continue;
+      }
 
-    int64_t fileSize;
-    rv = file->GetFileSize(&fileSize);
+      isEmpty = false;
+
+      int64_t fileSize;
+      rv = bodyFile->GetFileSize(&fileSize);
+      if (NS_WARN_IF(NS_FAILED(rv))) {
+        return rv;
+      }
+      MOZ_DIAGNOSTIC_ASSERT(fileSize >= 0);
+
+      aUsageInfo->AppendToFileUsage(Some(fileSize));
+    }
     if (NS_WARN_IF(NS_FAILED(rv))) {
       return rv;
     }
-    MOZ_DIAGNOSTIC_ASSERT(fileSize >= 0);
 
-    aUsageInfo->AppendToFileUsage(Some(fileSize));
+    if (isEmpty) {
+      QuotaInfo dummy;
+      mozilla::DebugOnly<nsresult> result =
+          RemoveNsIFileRecursively(dummy, bodyDir, /* aTrackQuota */ false);
+      // Try to remove the unexpected files, and keep moving on even if it
+      // fails because it might be created by virus or the operation system
+      MOZ_ASSERT(NS_SUCCEEDED(result));
+    }
+  }
+  if (NS_WARN_IF(NS_FAILED(rv))) {
+    return rv;
   }
 
   return NS_OK;
@@ -501,6 +547,9 @@
 
       NS_WARNING("Unknown Cache file found!");
     }
+    if (NS_WARN_IF(NS_FAILED(rv))) {
+      return rv;
+    }
 
     return NS_OK;
   }
